File: agent_framework/agentkit/subagents.py
"""Sub-Agents — ein Agent als Tool eines anderen.

Muster aus dem Notebook: Ein **Orchestrator** bekommt ein `delegate`-Tool. Jeder
Aufruf startet einen eigenständigen Agent-Loop (eigener Kontext, eigene Tools)
und gibt nur das **Ergebnis** zurück — Kontext-Isolation + Spezialisierung.

Zusammen mit `parallel_tools=True` laufen mehrere `delegate`-Aufrufe aus EINER
Orchestrator-Antwort nebenläufig (Map-Reduce / Supervisor).
"""

from __future__ import annotations

from typing import Optional

from .tools import ToolRegistry


def add_subagent(registry: ToolRegistry, name: str, description: str, llm,
                 tools: Optional[ToolRegistry] = None, system: Optional[str] = None,
                 strategy: str = "react", max_steps: int = 12,
                 param_name: str = "auftrag", param_desc: str = "Der Teilauftrag in Worten.",
                 parallel_tools: bool = True, bus=None) -> ToolRegistry:
    """Registriert einen Sub-Agenten als Tool im `registry` des Orchestrators.

    Jeder Aufruf erzeugt einen FRISCHEN Agent (eigenes Kurzzeitgedächtnis) und
    gibt dessen finale Antwort als Text zurück. Thread-safe, weil der
    OpenAI/Azure-Client thread-safe ist und jeder Sub-Agent eigenen State hat.

    Wird ein `bus` (EventBus) übergeben, werden ALLE Events des Sub-Agenten dorthin
    weitergeleitet — getaggt mit `source="<name>:<auftrag>"`, damit Consumer die
    (auch parallel laufenden) Sub-Agenten auseinanderhalten können.
    """
    # spät importieren -> kein Zirkelimport (agent.py nutzt nichts von hier)
    from .agent import Agent

    def _delegate(**kwargs):
        task = kwargs.get(param_name, "")
        sub = Agent(llm, tools=tools or ToolRegistry(), system=system,
                    strategy=strategy, max_steps=max_steps, parallel_tools=parallel_tools)
        if bus is None:
            return sub.run(task)
        # Event-Forwarding: Sub-Agent-Events laufen in den geteilten Bus.
        source = f"{name}:{' '.join(task.split())[:24]}"
        return sub.run_on_bus(task, bus, source=source)

    registry.add(
        name, description,
        {"type": "object",
         "properties": {param_name: {"type": "string", "description": param_desc}},
         "required": [param_name]},
        _delegate,
    )
    return registry

File: agent_framework_rs/src/subagents.rs
//! Sub-Agents — ein Agent als Tool eines anderen.
//!
//! Ein **Orchestrator** bekommt ein `delegate`-Tool (Name frei wählbar). Jeder
//! Aufruf baut über den normalen [`AgentBuilder`](crate::AgentBuilder) einen
//! FRISCHEN [`Agent`] — eigener Kontext, eigene Tools — führt ihn für den
//! Teilauftrag aus und gibt nur dessen **Ergebnis** zurück: Kontext-Isolation +
//! Spezialisierung. Zusammen mit `parallel_tools` laufen mehrere `delegate`-Aufrufe
//! aus EINER Antwort nebenläufig.
//!
//! Es gibt hier bewusst **keinen eigenen „Sub-Agent"-Typ**: ein Sub-Agent IST ein
//! [`Agent`], nur als Tool registriert. Diese Datei ist deshalb nur die dünne
//! Registrierung; Bau (über [`AgentBuilder`](crate::AgentBuilder)) und Lauf (über
//! [`Agent::run_as_subagent`]) teilt sie sich mit jedem anderen Agenten. Genau wie
//! Pythons `add_subagent` ist das eine Funktion, keine Klasse.

use crate::agent::{Agent, Strategy};
use crate::events::EventBus;
use crate::llm::Llm;
use crate::tools::ToolRegistry;
use serde_json::{json, Value};
use std::sync::Arc;

/// Registriert einen Sub-Agenten als Tool `name` im `registry` des Orchestrators.
///
/// Jeder Aufruf erzeugt einen FRISCHEN [`Agent`] (eigenes Kurzzeitgedächtnis) und
/// gibt dessen finale Antwort als Text zurück. Thread-safe, weil jeder Sub-Agent
/// eigenen State hat und der LLM-Client `Send + Sync` ist.
///
/// Wird ein `bus` übergeben, laufen ALLE Events des Sub-Agenten dorthin — getaggt
/// mit `source="<name>:<auftrag>"` (siehe [`Agent::run_as_subagent`]), damit
/// Consumer die (auch parallel laufenden) Sub-Agenten auseinanderhalten können.
///
/// Entspricht Pythons `add_subagent`: eine Funktion, kein eigener Builder-Typ.
#[allow(clippy::too_many_arguments)]
pub fn add_subagent(
    registry: &mut ToolRegistry,
    name: &str,
    description: &str,
    llm: Arc<dyn Llm>,
    tools: Option<ToolRegistry>,
    system: Option<&str>,
    strategy: Strategy,
    bus: Option<EventBus>,
) {
    let sub_name = name.to_string();
    let system = system.map(|s| s.to_string());
    let tools = tools.unwrap_or_default();
    let param_name = "auftrag";

    let params = json!({
        "type": "object",
        "properties": {param_name: {"type": "string", "description": "Der Teilauftrag in Worten."}},
        "required": [param_name],
    });

    registry.add(name, description, params, move |args: Value| {
        let task = args
            .get(param_name)
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();

        // Ein Sub-Agent ist ein ganz normaler Agent — gebaut über denselben Builder.
        let mut builder = Agent::builder(llm.clone())
            .tools(tools.clone())
            .strategy(strategy);
        if let Some(s) = &system {
            builder = builder.system(s);
        }
        let mut agent = builder.build();

        Ok(agent.run_as_subagent(&task, &sub_name, bus.as_ref(), None))
    });
}
